Add SCDF scaling overview, manual scaling recipe and autoscaling recipe (#172)
spring-io/dataflow.spring.io: content/documentation/pages/6-recipes/7-scaling/1-manual-scaling.md

---
path: 'recipes/scaling/manual-scaling/'
title: 'Manual Scaling'

> description: 'Using SCDF Shell for scaling applications'

Reviewer: > Scale applications using SCDF Shell

---

# Scale-out a streaming pipeline with SCDF shell

## Overview

> Deploy a streaming pipeline, simulate resources strangling scenario (like high CPU usage) and then use the SCDF shell to manually increase the number of strangled application instances to handle the increased load.

Reviewer: > Let's deploy a streaming pipeline, simulate a resource strangling scenario (e.g., high CPU spike) and then use the SCDF shell to manually increase the number of consumer application instances to handle the increased load.

![SCDF Manual Scaling](images/scdf-scale-manual.png)

> Instead of the SCDF Shell one can consider using the Scale REST API (explained in the next example) with some scripts or schedulers.

Reviewer: Remove this line. Confusing to see it here.

> This manual approach allows the TechOps teams to intervene and scale out when they observe high load. It also can be useful to scale the pipelines in case of an expected high load events. For example a retail store would pre-schedule scale out for events such as Black Friday and alike.

Reviewer: Remove TechOPs

Reviewer: Usually it is the developers in our case, at times, the Operators, as well.

> When more sophisticated approach is needed to control the scale-out and scale-in actions, one can leverage the SCDF monitoring and Prometheus builtin integration as shown in the follow-up example.

Reviewer: > When a more sophisticated approach is required to control the scale-out and scale-in operations in an automated manner, there's an option to leverage SCDF monitoring support with Prometheus.

## Prerequisite

> This recipe uses Kubernetes platform and Kafka middleware. Follow the [Kubectl](%currentPath%/installation/kubernetes/kubectl/) or the [Helm](%currentPath%/installation/kubernetes/helm/) installation instructions to setup Spring Cloud Data Flow with Kafka broker.

Reviewer: Not a recipe anymore

Reviewer: > This recipe uses the Kubernetes platform and Kafka as the messaging middleware. Follow the [Kubectl](%currentPath%/installation/kubernetes/kubectl/) or the [Helm](%currentPath%/installation/kubernetes/helm/) installation instructions to setup Spring Cloud Data Flow with Kafka broker.

For example the Helm installation would look like this:

```shell
helm install --name my-release stable/spring-cloud-data-flow --set kafka.enabled=true,rabbitmq.enabled=false,kafka.persistence.size=10Gi
```

And install the latest [kafka-docker](https://dataflow.spring.io/kafka-docker-latest) app starters.

> ## Scaling Recipe

Reviewer: Not a recipe

### Create stream pipeline

```shell
stream create --name scaletest --definition "time --fixed-delay=995 --time-unit=MILLISECONDS | transform --expression=\"payload + '-' + T(java.lang.Math).exp(700)\" | log"
```

> The `time` source generates timestamp messages on fixed time intervals (~ 1 msg/s), the `transform` processor performs a math operation to simulates a CPU intensive processing and the `log` sink prints the message payload.

Reviewer: > The `time` source generates current timestamp messages on a fixed time-interval (995ms = ~1msg/s), the `transform` processor performs a math operation to simulate a high CPU processing, and the `log` sink prints the transformed message payload.

### Deploy pipeline with data partitioning

```shell
stream deploy --name scaletest --properties "app.time.producer.partitionKeyExpression=payload,app.transform.spring.cloud.stream.kafka.binder.autoAddPartitions=true,app.transform.spring.cloud.stream.kafka.binder.minPartitionCount=4"
```

> The `producer.partitionKeyExpression=payload` property configures time’s output binding for partitioning. The partition key expression uses the message payload (e.g. the timestamp value) for computing the partitioned output channel.

Reviewer: > ... (e.g. the `toString()` value of the current timestamp) to compute how the data needs to be partitioned to the downstream output channels.

> The `spring.cloud.stream.kafka.binder.autoAddPartitions` deployment property instructs the Kafka binder to create new partitions when required. This is required if the topic’s partition size has not been pre configured.

Reviewer: > .. This is required if the topic is _not_ already over-partitioned

> The `spring.cloud.stream.kafka.binder.minPartitionCount` property sets the minimum number of partitions that the Kafka binder configures on topics on which the transformer app consumes data.

Reviewer: > The `spring.cloud.stream.kafka.binder.minPartitionCount` property sets the minimum number of partitions that the Kafka binder configures on the topic, which is where the transform-processor is subscribing for new data.

<!--TIP-->

When using RabbitMQ, the `app.log.spring.cloud.stream.kafka.binder.` deployment properties are not needed.

<!--END_TIP-->

Once the `scaletest` stream is deployed you should see:
![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-deployed.png)
Singe application instances for each application.

> Use the SCDF built in application Grafana dashboard to check the [stream application throughput metrics](%currentPath%/feature-guides/streams/monitoring/#prometheus-1):

Reviewer: built-in

![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-metrics.png)

The `time`, `transform` and `log` applications maintain the same message throughput (~ 1 message/s). The `transform` copes fine with the present load.

### Increase pipeline load

> Now lets increase the load by increasing the `time source` rate. By changing time's `time-unit` property from `MILLISECONDS` to `MICROSECONDS` the input rate will increase from one to couple of thousands messages per seconds.

Reviewer: > Now, let's increase the load by increasing the `time-source`'s message production rate. By changing time's `time-unit` property from `MILLISECONDS` to `MICROSECONDS` the input rate will increase from one to a couple of thousands of messages per second.

> Note that the [stream update](%currentPath%/stream-developer-guides/continuous-delivery/) functionality allows to update the time application without stopping the entire stream stream:

Reviewer: remove one "stream" in the end

```shell
stream update --name scaletest --properties "app.time.trigger.time-unit=MICROSECONDS"
```

> Now `time` emits messages with in a rate of `~ 5k` messages/s. The `transform` processor though is capped to around `1k msg/s` and in turn the capes throughput of the entire stream. Indicator that the `transform` has became a bottleneck.

Reviewer: > Now `time` source emits messages with a rate of ~5000 messages/s. The `transform` processor though is capped at around 1000 msg/s and that in turn, it halts the throughput of the entire stream to a certain level. This is an Indicator that the `transform` has become a bottleneck.

![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-bottleneck-metrics.png)

### Scale-Out

Use the SCDF Shell to scale the `transform` instances to 4:

```shell
stream scale app instances --name scaletest --applicationName transform --count 4
```

In result 3 additional transform instances are deployed:
![SCDF manual scaling scaletest deployed after scale out](images/scdf-scale-manual-scaletest-deployed-afer-scaleout.png)

> With the help of the additional instance the `log` sink and the entire pipeline catch up match with the `time` source:

Reviewer: > With the help of the additional instances of the `transform` processor, the entire data pipeline catches up to match with the `time` source's production rate.

![SCDF manual scaling scaletest metrics after scale out](images/scdf-scale-manual-scaletest-scaleout-metrics.png)

### Decrease pipeline load and Scale-In

If we reduce the input load back to the original rate 1 msg/s):

```shell
stream update --name scaletest --properties "app.time.trigger.time-unit=MILLISECONDS"
```

> the multiple `transform` instances gradually become redundant:

Reviewer: > The multiple `transform` processor instances aren't altering the overall throughput rate anymore. The extra unnecessary instances can be scaled back down:

---
path: 'recipes/scaling/manual-scaling/'
title: 'Manual Scaling'
description: 'Scale applications using SCDF Shell'
---

# Scale-out a streaming pipeline with SCDF shell

To learn about the basic scaling concepts in Spring Cloud Data Flow, please consult the [Scaling](%currentPath%/feature-guides/streams/scaling/) guide.

## Overview

Let's deploy a streaming pipeline, simulate a resource strangling scenario (e.g., high CPU spike) and then use the SCDF shell to manually increase the number of consumer application instances to handle the increased load.

![SCDF Manual Scaling](images/scdf-scale-manual.png)

> This manual approach allows the Operators to intervene and scale-out when they observe high load in a given application. For example, a retail store would want to pre-schedule a scale-out design for high-load and demanding days such as Black Friday.

Reviewer: ```suggestion
This manual approach allows the Operators and Developers to intervene and scale-out when they observe high-load in a given application. For example, a retail store may want to architect a scale-out design upfront to meet meet with the high-load and demanding days such as Black Friday.
```

> This manual approach allows the Operators and Developers to intervene and scale-out when they observe high-load in a given application. For example, a retail store may want to architect a scale-out design upfront to meet meet with the high-load and demanding days such as Black Friday.

Reviewer: I might have made a mistake in suggest edits. I see 2 "meet"s.

> When a more sophisticated approach is required to control the scale-out and scale-in operations in an automated manner, with the SCDF's monitoring support with Prometheus, users can configure and set it up for autoscaling. Check the [Autoscale streaming data pipelines with SCDF and Prometheus](%currentPath%/recipes/scaling/autoscaling/) recipe to learn how to do this.

Reviewer: ```suggestion
When a more sophisticated approach is required to control the scale-out and scale-in operations in an automated manner, with the help of SCDF's monitoring support using Prometheus, users can configure and set up autoscaling rules. Check the [Autoscale streaming data pipelines with SCDF and Prometheus](%currentPath%/recipes/scaling/autoscaling/) recipe to learn how to do this.
```

![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-bottleneck-metrics.png)

### Scale-Out

Use the SCDF Shell to scale the `transform` instances to 4:

```shell
stream scale app instances --name scaletest --applicationName transform --count 4
```

As a result of the previous command, 3 additional transform instances are deployed:

![SCDF manual scaling scaletest deployed after scale out](images/scdf-scale-manual-scaletest-deployed-afer-scaleout.png)

With the help of the additional instances of the `transform` processor, the entire data pipeline catches up to match with the `time` source's production rate.

![SCDF manual scaling scaletest metrics after scale out](images/scdf-scale-manual-scaletest-scaleout-metrics.png)

### Decrease pipeline load and Scale-In

If we reduce the source's data production rate back to the original rate (i.e., 1 msg/s):

```shell
stream update --name scaletest --properties "app.time.trigger.time-unit=MILLISECONDS"
```

> the extra `transform` processor instances aren't altering the overall throughput rate anymore. These extra instances can be scaled back in:

Reviewer: The

![SCDF manual scaling scaletest metrics redundant apps](images/scdf-scale-manual-scaletest-redundant-apps-metrics.png)

> allowing us to reduce the number of transform application instances to the original capacity:

Reviewer: Thus, allowing ..

---
path: 'recipes/scaling/manual-scaling/'
title: 'Manual Scaling'
description: 'Scale applications using SCDF Shell'
---

# Scale-out a streaming pipeline with SCDF shell

To learn about the basic scaling concepts in Spring Cloud Data Flow, please consult the [Scaling](%currentPath%/feature-guides/streams/scaling/) guide.

## Overview

Let's deploy a streaming pipeline, simulate a resource strangling scenario (e.g., high CPU spike) and then use the SCDF shell to manually increase the number of consumer application instances to handle the increased load.

![SCDF Manual Scaling](images/scdf-scale-manual.png)

> This manual approach allows the Operators to intervene and scale out when they observe high load. It also can be useful to scale the pipelines in case of an expected high load events. For example a retail store would pre-schedule scale out for events such as Black Friday and alike.

Reviewer: > This manual approach allows the Operators to intervene and scale-out when they observe high load in a given application. For example, a retail store would want to pre-schedule a scale-out design for high-load and demanding days such as Black Friday.

> When a more sophisticated approach is required to control the scale-out and scale-in operations in an automated manner, there's an option to leverage SCDF monitoring support with Prometheus. Check the [Autoscale streaming data pipelines with SCDF and Prometheus](%currentPath%/recipes/scaling/autoscaling/) recipe to learn how to do this.

Reviewer: > When a more sophisticated approach is required to control the scale-out and scale-in operations in an automated manner, with the SCDF's monitoring support with Prometheus, users can configure and set it up for autoscaling. Check the [Autoscale streaming data pipelines with SCDF and Prometheus](%currentPath%/recipes/scaling/autoscaling/) recipe to learn how to do this.

### Deploy pipeline with data partitioning

```shell
stream deploy --name scaletest --properties "app.time.producer.partitionKeyExpression=payload,app.transform.spring.cloud.stream.kafka.binder.autoAddPartitions=true,app.transform.spring.cloud.stream.kafka.binder.minPartitionCount=4"
```

The `producer.partitionKeyExpression=payload` property configures time’s output binding for partitioning. The partition key expression uses the message payload (e.g. the toString() value of the current timestamp) to compute how the data needs to be partitioned to the downstream output channels.
The `spring.cloud.stream.kafka.binder.autoAddPartitions` deployment property instructs the Kafka binder to create new partitions when required. This is required if the topic is not already over-partitioned.
The `spring.cloud.stream.kafka.binder.minPartitionCount` property sets the minimum number of partitions that the Kafka binder configures on the topic, which is where the transform-processor is subscribing for new data.

<!--TIP-->

> When using RabbitMQ, the `app.log.spring.cloud.stream.kafka.binder.*` deployment properties are not needed.

Reviewer: As suggested previously, let's remove any rabbit reference. This content is strictly only for SCDF + K8s + Kafka anyway.

<!--END_TIP-->

Once the `scaletest` stream is deployed you should see:
![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-deployed.png)
Singe application instances for each application.

> Use the SCDF built-in application Grafana dashboard to check the [stream application throughput metrics](%currentPath%/feature-guides/streams/monitoring/#prometheus-1):

Reviewer: > Use the SCDF's built-in Grafana dashboard to review the [stream application's throughput and the other application metrics](%currentPath%/feature-guides/streams/monitoring/#prometheus-1):

![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-metrics.png)

The `time`, `transform` and `log` applications maintain the same message throughput (~1 msg/s). The `transform` copes fine with the current load.

### Increase pipeline load

Now, let's increase the load by increasing the time-source's message production rate. By changing time's `time-unit` property from `MILLISECONDS` to `MICROSECONDS` the input rate will increase from one to a couple of thousands of messages per second.
Note that the [stream update](%currentPath%/stream-developer-guides/continuous-delivery/) functionality allows to rolling-update just the time application without stopping the entire stream:

```shell
stream update --name scaletest --properties "app.time.trigger.time-unit=MICROSECONDS"
```

> Now `time` source emits messages with a rate of `~5000 msg/s`. The `transform` processor though is capped at around `1000 msg/s` and that in turn, it halts the throughput of the entire stream to a certain level. This is an indicator that the `transform` has become a bottleneck.

Reviewer: > Now `time` source emits messages with a rate of `~5000 msg/s`. The `transform` processor, though, it is capped at around `1000 msg/s` and that in turn, it halts the throughput of the entire stream to a certain level. This is an indicator that the `transform` has become the bottleneck.

![SCDF manual scaling scaletest deployed](images/scdf-scale-manual-scaletest-bottleneck-metrics.png)

### Scale-Out

Use the SCDF Shell to scale the `transform` instances to 4:

```shell
stream scale app instances --name scaletest --applicationName transform --count 4
```

> In result 3 additional transform instances are deployed:

Reviewer: > As a result of the previous command, 3 additional transform instances are deployed:

![SCDF manual scaling scaletest deployed after scale out](images/scdf-scale-manual-scaletest-deployed-afer-scaleout.png)

With the help of the additional instances of the `transform` processor, the entire data pipeline catches up to match with the `time` source's production rate.

![SCDF manual scaling scaletest metrics after scale out](images/scdf-scale-manual-scaletest-scaleout-metrics.png)

### Decrease pipeline load and Scale-In

> If we reduce the input load back to the original rate (i.e., 1 msg/s):

Reviewer: > If we reduce the source's data production rate back to the original rate (i.e., 1 msg/s):

```shell
stream update --name scaletest --properties "app.time.trigger.time-unit=MILLISECONDS"
```

> The multiple `transform` processor instances aren't altering the overall throughput rate anymore. The extra unnecessary instances can be scaled back down:

Reviewer: > The extra `transform` processor instances aren't altering the overall throughput rate anymore. These extra instances can be scaled back down:
